src/Groups/matrices/linear_centralizer.jl

```diff
@@ -0,0 +1,426 @@
+   exp = fmpz(0)
+   for i in 1:length(L)-1, j in i+1:length(L)  exp += L[i][1]*L[i][2]*L[j][2] end
+   exp *= 2
+   exp += sum([(L[i][1]-1)*L[i][2]^2 for i in 1:length(L)])
+   exp *= degree(f)
+   res *= order(F)^exp
+
+   return listgens, res
+end
+
+# returns the list of generators
+function _centralizer_GL(x::MatElem)
+   _,cbm,ED = generalized_jordan_form(x; with_pol=true)    # cbm = change basis matrix
+   n=nrows(x)
+   listgens = MatElem[]
+   res = fmpz(1)
+   idN = identity_matrix(base_ring(x),n)
+
+   i=1
+   pos=1
+   f=ED[1][1]
+   V=[ED[1][2]]
+   while i <= length(ED)
+      if i<length(ED) && ED[i+1][1]==f
+         i+=1
+         push!(V,ED[i][2])
+      else
+         L = _centr_block_unipotent(f,base_ring(x),V)
+         for z in L[1]
+            push!(listgens, insert_block(idN,z,pos,pos))
+         end
+         res *= L[2]
+         pos += degree(f)*sum(V)
+         i+=1
+         if i<=length(ED)
+            f = ED[i][1]
+            V = [ED[i][2]]
+         end
+      end
+   end
+
+   return listgens, res, cbm
+end
+
+
+
+
+
+########################################################################
+#
+# Generators for SL
+#
+########################################################################
+
+# generators for GL(n,q) and SL(n,q) are described in:
+# Taylor, D. E., Pairs of Generators for Matrix Groups. I, The Cayley Bulletin, 3 (1987)
+
+# returns as matrices
+# returns a generating set for SL(n,F) of at most two elements
+function _gens_for_SL(n::Int, F::FinField)
+   n != 1 || return []
+   h1 = identity_matrix(F,n)
+   h2 = zero_matrix(F,n,n)
+   if order(F)==2 || order(F)==3
+      h1[1,2] = 1
+   else
+      h1[1,1] = primitive_element(F)
+      h1[2,2] = inv(h1[1,1])
+      h2[1,1] = -1
+   end
+   h2[1,n] = 1
+   for i in 1:n-1 h2[i+1,i] = -1 end
+   return h1,h2
+end
+
+# returns as matrices
+# does the same as above with the following changes:
+# the field F is replaced by F = F[x]/(f);
+# every entry y is replaced by a diagonal join of D copies of phi(y),
+# where phi is the ring automorphism sending a fixed root of f into the companion matrix of f.
+# (hence, the final output are still matrices with base ring F)
+# ASSUMPTION: deg(f) > 1
+function _gens_for_SL_matrix(f::PolyElem, n::Int, F::FinField; D::Int=1)
+   C = companion_matrix(f)
+   CP = evaluate(_centralizer(f),C)            # matrix of maximal order in the centralizer of the companion matrix
+   CPi = inv(CP)
+
+   n != 1 || return []
+   h1 = identity_matrix(F,n*degree(f)*D)
+   insert_block!(h1,diagonal_join([CP for i in 1:D]),1,1)
+   insert_block!(h1,diagonal_join([CPi for i in 1:D]),D*degree(f)+1,D*degree(f)+1)
+   h2 = zero_matrix(F,n*degree(f)*D,n*degree(f)*D)
+   for i in 1:(n-1)*degree(f)*D h2[i+degree(f)*D,i]=-1 end
+   for i in 1:degree(f)*D
+      h2[i,i]=-1
+      h2[i,i+(n-1)*degree(f)*D]=1
+   end
+# TODO waiting for a better solution, the h3 generator is necessary
```

Comment:
It is unclear to me what or whom this waiting for... Is this a research question? Then perhaps arXiv:2008.12651 discusses it, and we could point to that discussion here.

Or is this about some function or code missing? Then say what is missing, and how it should be used here once it's there.

Or is it something else? Again, please elaborate

Reply:
If `H` is an intermediate subgroup between `GL` and `SL`, we know how to generate `H` using three generators, and that's what I do in this piece of code. I'm much confident (if not sure) that it is possible to generate such `H` with only two generators, but I do not know how to do it at the moment. If in future we know how to generate such `H` for every finite field and dimension, the code can be improved by producing a smaller number of generator.
But it's an improvement from 3 to 2; if it is not worth it, I can just delete that line that could create confusion.

Reply:
So, if you included the explanation you just gave me here into that comment, that'd make me happy :-)